File: bindings/dotnet/src/Zlink/Contracts/Core/Zlink.cs
// SPDX-License-Identifier: MPL-2.0

using System;
using System.Collections.Generic;

namespace Systems.Zlink;

/// <summary>
/// Library entry point: factories for contexts, timers, pollers, and threads,
/// plus process-wide utilities (version, capabilities, proxying).
/// </summary>
public static class Zlink
{
    /// <summary>
    /// Raised when a user callback throws an exception. Callbacks run on a
    /// background dispatch thread, so such exceptions cannot propagate to the
    /// caller; subscribe here to observe them.
    /// </summary>
    public static event Action<Exception>? UnhandledCallbackException
    {
        add => CallbackExceptionHub.UnhandledCallbackException += value;
        remove => CallbackExceptionHub.UnhandledCallbackException -= value;
    }

    internal static int Errno()
    {
        return ZlinkRuntime.Errno();
    }

    /// <summary>
    /// Gets the human-readable message for a native zlink error code.
    /// </summary>
    public static string Strerror(int errnum)
    {
        return ZlinkRuntime.Strerror(errnum);
    }

    /// <summary>
    /// Gets the native zlink version.
    /// </summary>
    public static (int Major, int Minor, int Patch) Version()
    {
        return ZlinkRuntime.Version();
    }

    /// <summary>
    /// Creates a messaging context. The caller owns the returned context and
    /// must dispose it; disposing it terminates the sockets created from it.
    /// </summary>
    public static IContext CreateContext()
    {
        return new Context();
    }

    /// <summary>
    /// Creates an atomic counter. The caller owns the returned counter and must
    /// dispose it.
    /// </summary>
    public static IAtomicCounter CreateAtomicCounter()
    {
        return new AtomicCounter();
    }

    /// <summary>
    /// Creates a stopwatch. The caller owns the returned stopwatch and must
    /// dispose it.
    /// </summary>
    public static IZlinkStopwatch CreateStopwatch()
    {
        return new ZlinkStopwatch();
    }

    /// <summary>
    /// Creates and immediately starts a background thread running
    /// <paramref name="task"/>. The caller owns the returned thread and must
    /// dispose it (or <see cref="IZlinkThread.Join"/> then
    /// <see cref="IZlinkThread.Close"/>) to release it.
    /// </summary>
    public static IZlinkThread CreateThread(Action task)
    {
        return new ZlinkThread(task);
    }

    /// <summary>
    /// Creates a poller. The caller owns the returned poller and must dispose
    /// it.
    /// </summary>
    public static IPoller CreatePoller()
    {
        return new Poller();
    }

    /// <summary>
    /// Creates a standalone timer. The caller owns the returned timer and must
    /// dispose it.
    /// </summary>
    public static IZlinkTimer CreateTimer()
    {
        return new Timer();
    }

    /// <summary>
    /// Creates a timer bound to <paramref name="spot"/>'s event loop, so its
    /// callbacks run on the spot's dispatch thread. The caller owns the
    /// returned timer and must dispose it.
    /// </summary>
    public static IZlinkTimer CreateTimer(ISpot spot)
    {
        return Timer.FromSpot(SocketInterop.RequireSpot(spot, nameof(spot)));
    }

    /// <summary>
    /// Reports whether the native library was built with the named capability.
    /// </summary>
    public static bool Has(string capability)
    {
        return ZlinkRuntime.Has(capability);
    }

    /// <summary>
    /// Forwards messages between two sockets. This blocks the calling thread
    /// until the context is terminated, so run it on a dedicated thread.
    /// </summary>
    public static void Proxy(IZlinkSocket frontend, IZlinkSocket backend,
        IZlinkSocket? capture = null)
    {
        ZlinkRuntime.Proxy(frontend, backend, capture);
    }

    /// <summary>
    /// Forwards messages between two sockets under runtime control of
    /// <paramref name="control"/>. This blocks the calling thread until the
    /// proxy is terminated, so run it on a dedicated thread.
    /// </summary>
    public static void ProxySteerable(IZlinkSocket frontend, IZlinkSocket backend,
        IZlinkSocket? capture, IZlinkSocket control)
    {
        ZlinkRuntime.ProxySteerable(frontend, backend, capture, control);
    }

    /// <summary>
    /// Sleeps for the specified duration.
    /// </summary>
    public static void Sleep(TimeSpan duration)
    {
        ZlinkRuntime.Sleep(duration);
    }

    internal static void Sleep(int seconds)
    {
        ZlinkRuntime.Sleep(seconds);
    }

    /// <summary>
    /// Disposes each message in a multipart payload.
    /// </summary>
    public static void MultipartClose(IReadOnlyList<Message> parts)
    {
        ZlinkRuntime.MultipartClose(parts);
    }
}

File: bindings/java/src/main/java/systems/zlink/contracts/core/Zlink.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.core;

import systems.zlink.contracts.messaging.Message;
import systems.zlink.contracts.sockets.Socket;
import systems.zlink.runtime.nativeapi.InternalAccess;
import systems.zlink.runtime.nativeapi.Native;
import systems.zlink.runtime.nativeapi.NativeHelpers;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.time.Duration;
import java.util.List;
import java.util.Objects;

public final class Zlink {
    private Zlink() {}

    static int errno() {
        return Native.errno();
    }

    public static String strerror(int errnum) {
        return Native.strerror(errnum);
    }

    public static boolean has(String capability) {
        Objects.requireNonNull(capability, "capability");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment cap = NativeHelpers.toCString(arena, capability);
            return Native.has(cap) != 0;
        }
    }

    public static int[] version() {
        return Native.version();
    }

    public static void proxy(Socket frontend, Socket backend, Socket capture) {
        Objects.requireNonNull(frontend, "frontend");
        Objects.requireNonNull(backend, "backend");
        Native.proxy(InternalAccess.socketHandle(frontend),
            InternalAccess.socketHandle(backend),
            capture == null ? MemorySegment.NULL
                : InternalAccess.socketHandle(capture));
    }

    public static void proxySteerable(Socket frontend, Socket backend,
                                      Socket capture, Socket control) {
        Objects.requireNonNull(frontend, "frontend");
        Objects.requireNonNull(backend, "backend");
        Objects.requireNonNull(control, "control");
        Native.proxySteerable(InternalAccess.socketHandle(frontend),
            InternalAccess.socketHandle(backend),
            capture == null ? MemorySegment.NULL
                : InternalAccess.socketHandle(capture),
            InternalAccess.socketHandle(control));
    }

    static void sleep(int seconds) {
        if (seconds < 0)
            throw new IllegalArgumentException("seconds must be >= 0");
        Native.sleep(seconds);
    }

    public static void sleep(Duration duration) {
        Objects.requireNonNull(duration, "duration");
        long seconds = duration.toSeconds();
        if (seconds < 0 || seconds > Integer.MAX_VALUE) {
            throw new IllegalArgumentException(
              "duration seconds out of int range: " + seconds);
        }
        Native.sleep((int) seconds);
    }

    static void multipartClose(Message[] parts) {
        Message.closeAll(parts);
    }
}
